datafusion/functions-array/src/make_array.rs

```diff
@@ -131,6 +138,77 @@ impl ScalarUDFImpl for MakeArray {
     }
 }
 
+#[derive(Debug)]
+pub struct MakeArrayStrict {
```

Comment:
Can we just add the function that convert keys and values to list of expr instead of introducing another udf

Reply:
I think the high level idea is that


`SELECT MAKE_MAP('POST', 41, 'PAST', 33,'PATCH', 30)`
We arrange args to ['POST', 'PAST', 'PATCH'], [41, 33, 30], and call 
`MAP(['POST', 'PAST', 'PATCH'], [41, 33, 30])`

I just noticed that we can't directly pass these two array to `MapFunc` 😕

Reply:
I think we could figure how to build with dataframe API, map(keys, values) 

Current function is like
```rust
pub fn map($($arg: datafusion_expr::Expr),*) -> datafusion_expr::Expr {
  super::$FUNC().call(vec![$($arg),*])
}
```

Expected
```rust
pub fn map(keys: Vec<Expr>, values: Vec<Expr>) -> Expr
 ...
}
```

Reply:
For this PR, we just call `make_array_inner` and instead of `make_array_strict`, others we could deal with them in another PR

Reply:
I think we can't call `make_array_inner` in `ExprPlanner`. In my opinion, it should be used in the execution phase, but `ExprPlanner` is used in the planning phase. Our purpose is to arrange the input expressions into two array logical expressions (which is more efficient than arranging `ColumnarValue`).

However, I noticed that we don't have another way to create an array logical expression aside from `make_array`. We plan any array literal to `make_array`.

```
> explain select [c1,2] from t1;
+---------------+-----------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                    |
+---------------+-----------------------------------------------------------------------------------------+
| logical_plan  | Projection: make_array(CAST(t1.c1 AS Int64), Int64(2))                                  |
|               |   TableScan: t1 projection=[c1]                                                         |
| physical_plan | ProjectionExec: expr=[make_array(CAST(c1@0 AS Int64), 2) as make_array(t1.c1,Int64(2))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                         |
|               |                                                                                         |
+---------------+-----------------------------------------------------------------------------------------+
2 row(s) fetched. 
Elapsed 0.008 seconds.
```

Is there any other way to aggregate the expressions into an array or list in the planning phase?

Reply:
> I think we could figure how to build with dataframe API, map(keys, values)
> 

Sounds great. I think we can file another issue for it.

Reply:
I still think we should find a way to avoid `make_array_strict` 🤔

Reply:
We can change the MapFunc first, let it takes arguments with Vec<Expr>. The first half is `keys`, the other is `values`

Reply:
I play around to make sure the suggestion makes sense https://github.com/apache/datafusion/pull/11526

Reply:
Thanks! I will check it tonight.

Reply:
I have some concerns for it. If we make `MapFunc` to accept one array, it would be used like
```
SELECT map([1,2,3,'a','b','c'])
```
After planning, the input array would be `['1','2','3','a','b','c']` because of the type coercion for array elements. I think the behavior is wrong. If we change the signature of `MapFunc`, we might need to have another implementation to solve it.